packages/react-instantsearch-hooks/src/hooks/__tests__/useInstantSearch.test.tsx

```diff
@@ -0,0 +1,396 @@
+                  {children}
+                </>
+              ),
+            }),
+        }
+      );
+
+      // Initial render state from manual `getWidgetRenderState`
+      expect(result.current).toEqual(
+        expect.objectContaining({
+          uiState: {
+            indexName: { query: 'iphone' },
+          },
+          indexUiState: { query: 'iphone' },
+          setUiState: expect.any(Function),
+          setIndexUiState: expect.any(Function),
+        })
+      );
+
+      await waitForNextUpdate();
+
+      // InstantSearch.js state from the `render` lifecycle step
+      expect(result.current).toEqual(
+        expect.objectContaining({
+          uiState: {
+            indexName: { query: 'iphone' },
+          },
+          indexUiState: { query: 'iphone' },
+          setUiState: expect.any(Function),
+          setIndexUiState: expect.any(Function),
+        })
+      );
+    });
+
+    test('returns a function to modify the whole state', async () => {
+      function App() {
+        const { uiState, setUiState } = useInstantSearch();
+
+        return (
+          <>
+            <button
+              type="button"
+              onClick={() => {
+                setUiState({ indexName: { query: 'hey' } });
+              }}
+            >
+              {uiState.indexName.query}
+            </button>
+            <SearchBox />
+          </>
+        );
+      }
+
+      const { container } = render(
+        <InstantSearchHooksTestWrapper>
+          <App />
+        </InstantSearchHooksTestWrapper>
+      );
+
+      await wait(0);
+
+      expect(container.innerHTML).toMatchInlineSnapshot(
+        `"<button type=\\"button\\"></button>"`
+      );
+
+      userEvent.click(container.querySelector('button')!);
+
+      await wait(0);
+
+      expect(container.innerHTML).toMatchInlineSnapshot(
+        `"<button type=\\"button\\">hey</button>hey"`
+      );
+    });
+
+    test('returns a function to modify the index state', async () => {
+      function App() {
+        const { indexUiState, setIndexUiState } = useInstantSearch();
+
+        return (
+          <>
+            <button
+              type="button"
+              onClick={() => {
+                setIndexUiState({ query: 'hey' });
+              }}
+            >
+              {indexUiState.query}
+            </button>
+            <SearchBox />
+          </>
+        );
+      }
+
+      const { container } = render(
+        <InstantSearchHooksTestWrapper>
+          <App />
+        </InstantSearchHooksTestWrapper>
+      );
+
+      await wait(0);
+
+      expect(container.innerHTML).toMatchInlineSnapshot(
+        `"<button type=\\"button\\"></button>"`
+      );
+
+      userEvent.click(container.querySelector('button')!);
+
+      await wait(0);
+
+      expect(container.innerHTML).toMatchInlineSnapshot(
+        `"<button type=\\"button\\">hey</button>hey"`
+      );
+    });
+  });
+
+  describe('middleware', () => {
```

Comment:
Can we add a test with a real use case where you use `use()` in a component, to make sure that the middleware subscribes from the start, and doesn't miss any initial events?